scheduler/util.go

```diff
@@ -376,11 +377,29 @@ func taintedNodes(state State, allocs []*structs.Allocation) (map[string]*struct
 	return out, nil
 }
 
-// shuffleNodes randomizes the slice order with the Fisher-Yates algorithm
-func shuffleNodes(nodes []*structs.Node) {
+// shuffleNodes randomizes the slice order with the Fisher-Yates
+// algorithm. We seed the random source with the eval ID (which is
+// random) to aid in postmortem debuggging of specific evaluations and
+// state snapshots.
+func shuffleNodes(plan *structs.Plan, nodes []*structs.Node) {
+
+	// use the last 4 bytes because those are the random bits
+	// if we have sortable IDs
+	buf := []byte(plan.EvalID)
+	seed := int64(binary.BigEndian.Uint32(buf[len(buf)-4:]))
+
+	// include the plan progress so that we don't retry with
+	// the exact same shuffle
+	seed = seed |
+		int64(len(plan.NodeAllocation)) |
+		int64(len(plan.NodeUpdate)) |
+		int64(len(plan.DeploymentUpdates))
```

Comment:
If we place N allocs per attempt won't we end up with the same order? I'm still not sure that's a problem though. :thinking: 

Node reshuffling between attempts may workaround issues like #9506, and on attempts >1 these slices might all have 0 length resulting in not escaping a scheduler/planner disagreement. That's a lot of bad luck at once, and on the bright side this work + #11793 give us an incredible opportunity to fix otherwise intractable bugs.

So I'm **+1 shipping this as-is** if no one else can think of a more pathological condition _not_ reshuffling on retries could cause. If we are worried here are a couple options:

## Option 1: RefreshIndex

On partial application, the planner sets [`PlanResult.RefreshIndex`](https://github.com/hashicorp/nomad/blob/v1.2.5/nomad/plan_apply.go#L386-L392) which, in the absence of #9506 style bugs, should always increase between retries. Even with bugs like #9506 I would expect the Raft index to increase, but I don't think there's any guarantee.

`seed |= RefreshIndex` == `seed` when `RefreshIndex` is 0, so I think you could change this code to:

```suggestion
	// include the plan result's RefreshIndex so that we don't retry with
	// the exact same shuffle
	seed |= planResult.RefreshIndex
```

...once you plumb planResult all the way through... which might get ugly.

## Option 2: Retry Number
The retry number would be a nice as AFAICT it does not repeat for a specific eval, so `evalID ||  retry` should be unique per [`process()`](https://github.com/hashicorp/nomad/blob/v1.2.5/scheduler/generic_sched.go#L219) > [`computePlacements()`](https://github.com/hashicorp/nomad/blob/v1.2.5/scheduler/generic_sched.go#L473-L488) > [`shuffleNodes()`](https://github.com/hashicorp/nomad/blob/v1.2.5/scheduler/stack.go#L72-L73)

[`retryMax()`](https://github.com/hashicorp/nomad/blob/v1.2.5/scheduler/util.go#L315-L319) could be trivially altered to send the `attempts` int to the callback func (is `func()` now but would become `func(attempts int)`).

Adding a metric for attempts when `process()` succeeds would be trivial as well. I don't think we have a metric for `how many attempts is it taking evals to get processed?` and if it's >1 attempt that's an indication of scheduler contention. If it's occurring persistently operators may want to lower `num_schedulers` as they're just throwing CPU cycles away with plans that end up conflicting.

Reply:
I really like option 1 here. It shouldn't be too hard to plumb that through and it's a better indicator than we'd get from counts, especially because of:

> and on attempts >1 these slices might all have 0 length resulting in not escaping a scheduler/planner disagreement. 

> Adding a metric for attempts when `process()` succeeds would be trivial as well. I don't think we have a metric for `how many attempts is it taking evals to get processed?` and if it's >1 attempt that's an indication of scheduler contention.

Hm... I think that can be right at the top of `process`? Would we do that as a Sample type so that operators can get a histogram over it?